lib/tcx.js

```diff
@@ -0,0 +1,207 @@
+  ["TotalTimeSeconds", "totalTimeSeconds"],
+  ["DistanceMeters", "distanceMeters"],
+  ["MaximumSpeed", "maxSpeed"],
+  ["AverageHeartRateBpm", "avgHeartRate"],
+  ["MaximumHeartRateBpm", "maxHeartRate"],
+  // Extended Lap attributes
+  ["AvgSpeed", "avgSpeed"],
+  ["AvgWatts", "avgWatts"],
+  ["MaxWatts", "maxWatts"],
+];
+
+function getProperties(node) {
+  const prop = {};
+  const extensions = node.getElementsByTagNameNS(
+    "http://www.garmin.com/xmlschemas/ActivityExtension/v2",
+    "*"
+  );
+  attributeNames.forEach((attr) => {
+    const raw = get1(node, attr[0]);
+    if (raw !== null) {
+      prop[attr[1]] = parseFloat(nodeVal(raw));
+    } else if (extensions !== null) {
+      for (let i = 0; i < extensions.length; i++) {
+        const extension = extensions[i];
+        if (extension.localName === attr[0]) {
+          const ext = nodeVal(extension);
+          if (ext !== null) {
+            const v = parseFloat(ext);
+            if (!isNaN(v)) {
+              prop[attr[1]] = v;
+            }
+          }
+        }
+      }
+    }
+  });
+  return prop;
+}
+
+function coordPair(x) {
+  const lon = nodeVal(get1(x, "LongitudeDegrees"));
+  const lat = nodeVal(get1(x, "LatitudeDegrees"));
+  if (!lon.length || !lat.length) {
+    return null;
+  }
+  const ll = [parseFloat(lon), parseFloat(lat)];
+  const alt = get1(x, "AltitudeMeters");
+  const heartRate = get1(x, "HeartRateBpm");
+  const cadence = get1(x, "Cadence");
+  const time = get1(x, "Time");
+  let a;
+  if (alt) {
+    a = parseFloat(nodeVal(alt));
+    if (!isNaN(a)) {
+      ll.push(a);
+    }
+  }
+  const result = {
+    coordinates: ll,
+    time: time ? nodeVal(time) : null,
+    heartRate: heartRate ? parseFloat(nodeVal(heartRate)) : null,
+    cadence: cadence ? parseFloat(nodeVal(cadence)) : null,
+  };
+  const extensions = x.getElementsByTagNameNS(
+    "http://www.garmin.com/xmlschemas/ActivityExtension/v2",
+    "*"
+  );
+  if (extensions !== null) {
+    for (let i = 0; i < extensions.length; i++) {
+      const extension = extensions[i];
+      attributeNames
+        .map((r) => r[0])
+        .filter((n) => n === extension.localName)
+        .forEach((name) => {
+          const raw = nodeVal(extension);
+          if (raw !== null) {
+            const v = parseFloat(raw);
+            if (!isNaN(v)) {
+              result[name] = v;
+            }
+          }
+        });
+    }
+  }
+  return result;
+}
+
+function getPoints(node, pointname) {
+  const pts = node.getElementsByTagName(pointname);
+  const line = [];
+  const times = [];
+  const l = pts.length;
+  const extendedValues = {};
+  if (l < 2) return {}; // Invalid line in GeoJSON
+  for (let i = 0; i < l; i++) {
+    const c = coordPair(pts[i]);
+    if (c === null) continue;
+    line.push(c.coordinates);
+    if (c.time) times.push(c.time);
+    attributeNames
+      .map((r) => r[0])
+      .forEach((name) => {
+        if (c[name] || extendedValues[name]) {
```

Comment:
Looping over all potential attributes here seems wasteful (esp when done with .map, which will create a temporary array per iteration) - would be better to structure the coordinate output in a way that you could loop over the coordinate's extra properties.